agent/exec/controller.go

```diff
@@ -265,65 +197,68 @@ func Do(ctx context.Context, task *api.Task, ctlr Controller) (*api.TaskStatus,
-	case api.TaskStateNew, api.TaskStateAllocated,
-		api.TaskStateAssigned, api.TaskStateAccepted,
-		api.TaskStatePreparing, api.TaskStateReady,
-		api.TaskStateStarting, api.TaskStateRunning,
-		api.TaskStateCompleted, api.TaskStateFailed,
-		api.TaskStateRejected:
-
-		if task.DesiredState < status.State {
-			// do not yet proceed. the desired state is less than the current
-			// state.
+	if task.DesiredState == api.TaskStateShutdown {
+		if status.State >= api.TaskStateCompleted {
 			return noop()
 		}
 
-		switch status.State {
-		case api.TaskStateNew, api.TaskStateAllocated,
-			api.TaskStateAssigned:
-			return transition(api.TaskStateAccepted, "accepted")
-		case api.TaskStateAccepted:
-			return transition(api.TaskStatePreparing, "preparing")
-		case api.TaskStatePreparing:
-			if err := ctlr.Prepare(ctx); err != nil && err != ErrTaskPrepared {
-				return fatal(err)
-			}
+		if err := ctlr.Shutdown(ctx); err != nil {
+			return fatal(err)
+		}
 
-			return transition(api.TaskStateReady, "prepared")
-		case api.TaskStateReady:
-			return transition(api.TaskStateStarting, "starting")
-		case api.TaskStateStarting:
-			if err := ctlr.Start(ctx); err != nil && err != ErrTaskStarted {
-				return fatal(err)
-			}
+		return transition(api.TaskStateShutdown, "shutdown")
+	}
 
-			return transition(api.TaskStateRunning, "started")
-		case api.TaskStateRunning:
-			if err := ctlr.Wait(ctx); err != nil {
-				if _, ok := err.(*ExitError); ok {
-					return transition(api.TaskStateFailed, "failed")
-				}
-
-				// TODO(stevvooe): In most cases, failures of Wait are actually
-				// not a failure of the task. We account for this in fatal by
-				// checking temporary.
-				return fatal(err)
-			}
+	if status.State > task.DesiredState {
+		return noop() // way beyond desired state, pause
+	}
 
-			return transition(api.TaskStateCompleted, "finished")
-		}
-	case api.TaskStateShutdown:
-		if status.State >= api.TaskStateShutdown {
-			return noop()
+	// the following states may proceed past desired state.
+	switch status.State {
+	case api.TaskStatePreparing:
+		if err := ctlr.Prepare(ctx); err != nil && err != ErrTaskPrepared {
+			return fatal(err)
 		}
 
-		if err := ctlr.Shutdown(ctx); err != nil {
+		return transition(api.TaskStateReady, "prepared")
+	case api.TaskStateStarting:
+		if err := ctlr.Start(ctx); err != nil && err != ErrTaskStarted {
 			return fatal(err)
 		}
 
-		return transition(api.TaskStateShutdown, "shutdown")
+		return transition(api.TaskStateRunning, "started")
+	case api.TaskStateRunning:
+		if err := ctlr.Wait(ctx); err != nil {
+			// Wait should only proceed to failed if there is a terminal
+			// error. The only two conditions when this happens are when we
+			// get an exit code or when the container doesn't exist.
+			switch err := err.(type) {
+			case ExitCoder:
+				return transition(api.TaskStateFailed, "failed")
+			default:
+				// pursuant to the above comment, report fatal, but wrap as
```

Comment:
`The only two conditions when this happens are when we get an exit code or when the container doesn't exist.`
@stevvooe Are we able to handle `container doesn't exist` error correctly here? 
I got the same problem described in [docker-1.12-integration/issues/401](https://github.com/docker/docker-1.12-integration/issues/401) and found that in `default` branch we got `no such container` error.